aten/src/ATen/native/cpu/ReduceOpsKernel.cpp

```diff
@@ -1,118 +1,154 @@
 #include "ATen/native/cpu/ReduceOpsKernel.h"
+
+#include <numeric>
+
 #include "ATen/Dispatch.h"
 #include "ATen/Parallel.h"
+#include "ATen/optional.h"
+#include "ATen/cpu/vec256/vec256.h"
 
 namespace at {
 namespace native {
 
 using namespace vec256;
 
-// This adds the content of arr to sum
-template <class scalar_t, template <class> class OP, CPUCapability C>
-inline scalar_t allreduce_kernel_(const scalar_t *arr, size_t start, size_t end,
-                                  scalar_t sum) {
-  Vec256<scalar_t> part_sum;
-  // Use all 16 registers.
-  Vec256<scalar_t> tmp_sum[4], tmp_sum1, tmp_sum2, tmp_sum3;
-  Vec256<scalar_t> a[8];
-  size_t width =
-      256 / sizeof(scalar_t); // primitives per 256 bytes (two cache lines)
-  size_t epr = 32 / sizeof(scalar_t); // primitives per Vec256
-  size_t k = 0;
-  for (; k < (end - start) / width; k++) {
-    for (size_t i = 0; i < 8; i++) {
-      a[i].load(arr + (k * width) + i * epr + start);
+static inline int64_t round_down(int64_t a, int64_t m) {
+  return a - (a % m);
+}
+
+template<typename F>
+static void parallel_for(int64_t end, int64_t step, bool parallelize, F func) {
```

Comment:
Why not put this into parallel.h?

Reply:
init_tbb_num_threads is called by Reduction::apply which is the only entry point.

It's not in parallel.h because I'm not sure this pattern is generally useful as-is. If we start using it in another file then we can move it to a common header.